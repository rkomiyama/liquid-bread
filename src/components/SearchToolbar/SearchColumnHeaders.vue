<template>
  <v-list dense>
    <v-subheader>Column headers</v-subheader>
    <v-list-tile v-for="(header, i) in headers" :key="i">
      <v-list-tile-action>
        <v-checkbox v-model="checked[i]"></v-checkbox>
      </v-list-tile-action>
      <v-list-tile-content>
        <v-list-tile-title>{{ header.text }}</v-list-tile-title>
      </v-list-tile-content>
    </v-list-tile>
  </v-list>
</template>

<script>
export default {
  name: "SearchColumnHeaders",
  data() {
    return {
      checked: []
    };
  },
  created() {
    this.checked = this.passedChecks;
  },
  watch: {
    checked(checked) {
      if (checked && checked.length) {
        this.$emit("change:checks", checked);
      }
    }
  },
  props: {
    headers: {
      type: Array,
      default: () => []
    },
    passedChecks: {
      type: Array,
      default: () => []
    }
  }
};
</script>

<style lang="scss" scoped>
</style>
