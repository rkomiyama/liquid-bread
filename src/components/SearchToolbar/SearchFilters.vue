<template>
  <v-card>
    <v-subheader>Search filters</v-subheader>
    <v-card-text class="pt-0">
      <v-layout wrap justify-space-between>
        <v-flex lg3 md3 sm5>
          <v-text-field v-model="filters.abvLt" label="ABV (less than)" hide-details></v-text-field>
          <v-text-field v-model="filters.abvGt" label="ABV (greater than)" hide-details></v-text-field>
          <v-text-field v-model="filters.ibuLt" label="IBU (less than)" hide-details></v-text-field>
          <v-text-field v-model="filters.ibuGt" label="IBU (greater than)" hide-details></v-text-field>
          <v-text-field v-model="filters.ebcLt" label="EBC (less than)" hide-details></v-text-field>
          <v-text-field v-model="filters.ebcGt" label="EBC (greater than)" hide-details></v-text-field>
        </v-flex>
        <v-flex lg3 md3 sm5>
          <v-text-field v-model="filters.beerName" label="Beer name" hide-details></v-text-field>
          <v-text-field v-model="filters.yeastName" label="Yeast" hide-details></v-text-field>
          <v-text-field v-model="filters.hopsName" label="Hops" hide-details></v-text-field>
          <v-text-field v-model="filters.maltName" label="Malt" hide-details></v-text-field>
          <v-text-field v-model="filters.foodName" label="Food" hide-details></v-text-field>
          <v-text-field v-model="filters.ids" label="IDs" hide-details></v-text-field>
        </v-flex>
        <v-flex lg3 md3 sm12>
          <v-menu
            ref="beforeDatepicker"
            :close-on-content-click="false"
            v-model="beforeDatepicker"
            :nudge-right="40"
            :return-value.sync="filters.beforeDate"
            lazy
            transition="scale-transition"
            offset-y
            full-width
            min-width="250px"
          >
            <v-text-field
              slot="activator"
              v-model="filters.beforeDate"
              label="Brewed before"
              prepend-icon="event"
              readonly
              clearable
              hide-details
            ></v-text-field>
            <v-date-picker
              v-model="filters.beforeDate"
              @input="$refs.beforeDatepicker.save(filters.beforeDate)"
              type="month"
            ></v-date-picker>
          </v-menu>
          <v-menu
            ref="afterDatepicker"
            :close-on-content-click="false"
            v-model="afterDatepicker"
            :nudge-right="40"
            :return-value.sync="filters.afterDate"
            lazy
            transition="scale-transition"
            offset-y
            full-width
            min-width="250px"
          >
            <v-text-field
              slot="activator"
              v-model="filters.afterDate"
              label="Brewed after"
              prepend-icon="event"
              readonly
              clearable
              hide-details
            ></v-text-field>
            <v-date-picker
              v-model="filters.afterDate"
              @input="$refs.afterDatepicker.save(filters.afterDate)"
              type="month"
            ></v-date-picker>
          </v-menu>
        </v-flex>
      </v-layout>
    </v-card-text>
  </v-card>
</template>

<script>
/**
 * Component containing the search filters
 */
export default {
  name: "SearchFilters",
  props: {
    /**
     * Flag to check if filters have been applied
     */
    applyFilters: {
      type: Boolean
    },
    /**
     * Flag to check if filters need to be cleared
     */
    clearFilters: {
      type: Boolean
    }
  },
  watch: {
    applyFilters(val) {
      if (val) {
        /**
         * Triggered when the search button is clicked
         * @event submit:applySearchFilters
         * @type {Object}
         */
        this.$emit("submit:applySearchFilters", { ...this.filters });
      }
    },
    clearFilters(val) {
      if (val) {
        this.filters = {
          abvLt: null,
          abvGt: null,
          ibuLt: null,
          ibuGt: null,
          ebcLt: null,
          ebcGt: null,
          beerName: "",
          yeastName: "",
          beforeDate: null,
          afterDate: null,
          hopsName: "",
          maltName: "",
          foodName: "",
          ids: ""
        };
        this.beforeDatepicker = null;
        this.afterDatepicker = null;
      }
    }
  },
  data() {
    return {
      filters: {
        abvLt: null,
        abvGt: null,
        ibuLt: null,
        ibuGt: null,
        ebcLt: null,
        ebcGt: null,
        beerName: "",
        yeastName: "",
        beforeDate: null,
        afterDate: null,
        hopsName: "",
        maltName: "",
        foodName: "",
        ids: ""
      },
      beforeDatepicker: null,
      afterDatepicker: null
    };
  }
};
</script>

<style lang="scss" scoped>
</style>

<docs>
```jsx
<SearchFilters :style="{height: '200px', overflowY: 'scroll'}" />
```
</docs>