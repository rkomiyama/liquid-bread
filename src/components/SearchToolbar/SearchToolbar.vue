<template>
  <v-toolbar app ref="toolbar" :extension-height="height">
    <v-toolbar-title>Liquid Bread</v-toolbar-title>
    <v-spacer></v-spacer>
    <v-toolbar-side-icon @click="showMore = !showMore"></v-toolbar-side-icon>
    <v-layout v-if="showMore" slot="extension">
      <v-flex xs8>
        <SearchFilters
          class="filters-list"
          @submit:applySearchFilters="filtersChangeHandler"
          :applyFilters="applyFilters"
        />
        <v-btn @click.native="applyFilters = true">Search</v-btn>
      </v-flex>
      <v-spacer></v-spacer>
      <SearchColumnHeaders
        class="headers-list"
        @change:checks="checksChangeHandler"
        :headers="headers"
        :passedChecks="checked"
      />
    </v-layout>
  </v-toolbar>
</template>

<script>
import SearchFilters from "./SearchFilters";
import SearchColumnHeaders from "./SearchColumnHeaders";

export default {
  name: "SearchToolbar",
  data() {
    return {
      showMore: false,
      applyFilters: false,
      headers: [
        { text: "ID", value: "id", width: "50px" },
        { text: "Name", value: "name", width: "35%" },
        { text: "ABV", value: "abv", width: "50px" },
        { text: "IBU", value: "ibu", width: "50px" },
        { text: "EBC", value: "ebc", width: "50px" },
        { text: "SRM", value: "srm", width: "50px" },
        { text: "First brewed", value: "first_brewed" },
        { text: "Volume", value: "volume" },
        {
          text: "Atten. level",
          value: "attenuation_level",
          width: "40px"
        },
        { text: "pH", value: "ph", width: "50px" },
        { text: "Boil volume", value: "boil_volume", width: "50px" },
        { text: "Target OG", value: "target_og", width: "50px" },
        { text: "Target FG", value: "target_fg", width: "50px" }
      ],
      checked: [],
      height: undefined
    };
  },
  mounted() {
    this.showMore = true;
  },
  created() {
    this.checked = Array(this.headers.length).fill(false);
    this.checked.fill(true, 0, 7);
  },
  watch: {
    showMore(val) {
      if (val) {
        this.$nextTick(function() {
          this.height = this.$refs.toolbar.$el.lastChild.firstChild.clientHeight;
        });
      }
    },
    applyFilters(val) {
      if (val) {
        this.$nextTick(function() {
          this.applyFilters = false;
        });
      }
    }
  },
  methods: {
    /**
     * Gets called when filters change
     * @param {Object} filters Filters object
     * @public
     */
    filtersChangeHandler(filters) {
      /**
       * Triggered when the filters have changed
       * @event change:filters
       * @type {Object}
       */
      this.$emit("change:filters", filters);
    },
    /**
     * Gets called when checkboxes change
     * @param {Array} checks Checks array
     * @public
     */
    checksChangeHandler(checks) {
      const selectedHeaders = this.headers.filter((header, i) => checks[i]);
      /**
       * Triggered when the headers selection has changed
       * @event change:headers
       * @type {Object}
       */
      this.$emit("change:headers", selectedHeaders);
      this.checked = checks;
    }
  },
  components: {
    SearchColumnHeaders,
    SearchFilters
  }
};
</script>

<style lang="scss" scoped>
.headers-list,
.filters-list {
  height: 200px;
  overflow-y: scroll;
}
</style>

<docs>
```jsx
<SearchToolbar :style="{position: 'relative'}" />
```
</docs>