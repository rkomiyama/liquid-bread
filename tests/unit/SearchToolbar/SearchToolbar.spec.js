import Vue from "vue";
import Vuetify from "vuetify";
import { shallowMount } from "@vue/test-utils";
import { createRenderer } from "vue-server-renderer";
import SearchToolbar from "@/components/SearchToolbar/SearchToolbar.vue";

Vue.use(Vuetify);

let wrapper;

describe("SearchToolbar.vue", () => {
  beforeEach(() => {
    wrapper = shallowMount(SearchToolbar);
  });

  it("changes height when showMore is changed", () => {
    const height = wrapper.vm.height;
    expect(wrapper.vm.showMore).toBeTruthy();
    wrapper.setData({ showMore: false, height: 100 });
    expect(wrapper.vm.showMore).toBeFalsy();
    expect(wrapper.vm.height).not.toEqual(height);
  });

  it("applies filters", done => {
    expect(wrapper.vm.applyFilters).toBeFalsy();
    const searchButton = wrapper.find("v-btn-stub[type='button']");
    searchButton.trigger("click");
    expect(wrapper.vm.applyFilters).toBeTruthy();
    Vue.nextTick(() => {
      expect(wrapper.vm.applyFilters).toBeFalsy();
      done();
    });
  });

  it("calls filtersChangeHandler", () => {
    expect(wrapper.emitted("change:filters")).toBeFalsy();
    wrapper.vm.filtersChangeHandler();
    expect(wrapper.emitted("change:filters")).toBeTruthy();
  });

  it("calls checksChangeHandler", () => {
    let checks = Array(wrapper.vm.headers.length).fill(false);
    expect(wrapper.emitted("change:headers")).toBeFalsy();
    wrapper.vm.checksChangeHandler(checks);
    expect(wrapper.emitted("change:headers")).toBeTruthy();
  });

  it("SearchToolbar has same HTML structure", () => {
    const renderer = createRenderer();
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });
});
