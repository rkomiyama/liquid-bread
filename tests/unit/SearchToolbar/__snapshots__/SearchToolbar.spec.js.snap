// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchToolbar.vue SearchToolbar has same HTML structure 1`] = `
<v-toolbar-stub app="true" extensionHeight="0" scrollThreshold="300">
  <v-toolbar-title-stub>Liquid Bread</v-toolbar-title-stub>
  <v-spacer-stub></v-spacer-stub>
  <v-toolbar-side-icon-stub></v-toolbar-side-icon-stub>
  <v-layout-stub tag="div">
    <v-flex-stub tag="div" xs8="">
      <SearchFilters-stub class="filters-list"></SearchFilters-stub>
      <v-btn-stub activeClass="v-btn--active" tag="button" type="button">Search</v-btn-stub>
    </v-flex-stub>
    <v-spacer-stub></v-spacer-stub>
    <SearchColumnHeaders-stub headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" passedChecks="true,true,true,true,true,true,true,false,false,false,false,false,false" class="headers-list"></SearchColumnHeaders-stub>
  </v-layout-stub>
</v-toolbar-stub>
`;
